### DISCUSSION TITLE
CDT for 3D surface meshes intersection

### DISCUSSION THREAD

**Discussion by VinceChiarut** [artem-ogre/CDT#137] [General]
A little message to thank the CDT developers. I am currently developing a mesh backend for a FE solver using intel TBB and I needed an opensource constrained Delaunay kernel for surface meshing. I was very pleased with the easy interface offered by CDT and its very good performance for my usage. I have developed a triangle-surface mesh intersection process for now. This approach is performed in two stages:
1. A geometric triangle/triangle intersection stage with multi-threading and BVH, delivering a thread-safe conformal segments/vertices data structure.
2. The embarrassingly parallel conforming remeshing of each of the intersected triangles using CDT.

Here is an example of an academic application: 6859 intersecting spheres with 8.7 Mtriangles. The conform intersection (17.2 Mtriangles, 110 for the worst triangle quality) was obtained in 11.7s with 6 threads and 36.6s with 1 thread on a 6 core laptop (21GB of memory required during the operations).

![Capture d’écran du 2023-09-19 22-27-35](https://github.com/artem-ogre/CDT/assets/142824348/aa7ab767-03f1-40c0-88e0-b72c56e8ad7c)
![Capture d’écran du 2023-09-19 22-29-06](https://github.com/artem-ogre/CDT/assets/142824348/bc4d376c-105b-4a5f-ae70-f1967b1a808c)

Once again, I extend my deepest appreciation to the creators of CDT for their excellent open source contribution.

**Comment by artem-ogre** [comment]
Hi @VinceChiarut, 

Thank you very much for the kind words! This made my day :) It was also very interesting to know about your CDT usage. 

Do you also use robust predicates for detecting or resolving triangle-triangle intersections?

**Comment by VinceChiarut** [comment]
Hi @artem-ogre,
For my application I use a global tolerance to identify singular case (edge/edge, edge/vertex or vertex/plane) intersections in the geometric part. It's not perfect, but I works pretty well.
